src/Microsoft.AspNet.Buffering/BufferingWriteStream.cs

```diff
@@ -0,0 +1,186 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Buffering
+{
+    internal class BufferingWriteStream : Stream
+    {
+        private readonly Stream _innerStream;
+        private readonly MemoryStream _buffer = new MemoryStream();
+        private bool _isBuffering = true;
+
+        public BufferingWriteStream(Stream innerStream)
+        {
+            _innerStream = innerStream;
+        }
+
+        public override bool CanRead
+        {
+            get { return false; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return _isBuffering || _innerStream.CanSeek; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return _innerStream.CanWrite; }
+        }
+
+        public override long Length
+        {
+            get
+            {
+                if (_isBuffering)
```

Comment:
Hate to say it, but it's probably a good idea to maintain a `long _length` in this class that's adjusted when Write's, Seek, set Position, and SendFile are called... That way the behavior of the stream's Position and Length props can be deterministic and non-throwing in all cases...

It should also probably throw an exception all of the time when _isBuffering == false when someone calls Seek or set Position